crud/src/main/java/org/openmbee/sdvc/crud/services/DefaultNodeService.java

```diff
@@ -130,6 +137,8 @@ public ElementsResponse createOrUpdate(String projectId, String refId, ElementsR
         ContextHolder.setContext(projectId, refId);
         boolean overwriteJson = Boolean.parseBoolean(params.get("overwrite"));
 
+        preCommitChanges(projectId, refId, req.getElements(), params, user);
```

Comment:
Since the artifact subscriber is relying on checking for the artifact controller, this feels like it should really be called at the controller level - it'll basically be a hook for processing input going into /elements - the criteria for applying the processing is url based. The service interface also isn't guaranteed to be invoked through a request handler.

It may be worthwhile to also have a hook for dealing with input before it gets processed at the service level, but I think that would be something along the lines of a subclass override.

Reply:
Reworked this into a generic EmbeddedHookService, and installed the hook at the controller level.